File: export-candidates.mjs
#!/usr/bin/env node
/**
 * export-candidates.mjs — Convert mined candidates to ontology submission formats
 *
 * Reads a candidates JSON file produced by ontology-miner.mjs or ontology-agent.mjs
 * and exports submission-ready files for ontology curators.
 *
 * Output formats:
 *   obo      — OBO stanza blocks, ready to paste into a .obo file or GitHub PR
 *   robot    — ROBOT template TSV, importable with `robot template`
 *   github   — GitHub New Term Request (NTR) markdown, one file per term
 *   csv      — Flat CSV for spreadsheet-based human review
 *   all      — All of the above (default)
 *
 * Usage:
 *   node export-candidates.mjs --input ontologies/candidates/my-paper-2026-06-05.json
 *   node export-candidates.mjs --input ontologies/candidates/my-paper-2026-06-05.json --format obo
 *   node export-candidates.mjs --input ontologies/candidates/my-paper-2026-06-05.json --format github
 *   node export-candidates.mjs --list   (show all available candidate files)
 *
 * Submission targets:
 *   PO  → https://github.com/Planteome/plant-ontology/issues/new
 *   TO  → https://github.com/Planteome/plant-trait-ontology/issues/new
 *   PECO→ https://github.com/Planteome/plant-experimental-conditions-ontology/issues/new
 */

import { readFileSync, writeFileSync, mkdirSync, readdirSync, existsSync } from 'fs';
import { join, dirname, basename, extname } from 'path';
import { fileURLToPath } from 'url';

const ROOT    = dirname(fileURLToPath(import.meta.url));
const CAND_DIR = join(ROOT, 'ontologies', 'candidates');
const EXP_DIR  = join(ROOT, 'ontologies', 'exports');

// ---------------------------------------------------------------------------
// Args
// ---------------------------------------------------------------------------

const args = process.argv.slice(2);
function getArg(flag) {
  const i = args.indexOf(flag);
  const v = i !== -1 ? args[i + 1] : undefined;
  return v && !v.startsWith('--') ? v : null;
}

if (args.includes('--list')) {
  if (!existsSync(CAND_DIR)) { console.log('No candidates directory found.'); process.exit(0); }
  const files = readdirSync(CAND_DIR).filter(f => f.endsWith('.json') && f !== '.gitkeep');
  if (!files.length) { console.log('No candidate files found in ontologies/candidates/'); process.exit(0); }
  console.log(`Candidate files in ontologies/candidates/:\n`);
  for (const f of files.sort().reverse()) {
    const data = JSON.parse(readFileSync(join(CAND_DIR, f), 'utf8'));
    console.log(`  ${f}  (${data.candidate_count ?? 0} candidates)`);
  }
  process.exit(0);
}

const inputArg  = getArg('--input');
const formatArg = getArg('--format') ?? 'all';

if (!inputArg) {
  console.error('Usage: node export-candidates.mjs --input <candidates-json> [--format obo|robot|github|csv|all]');
  console.error('       node export-candidates.mjs --list');
  process.exit(1);
}

const inputPath = inputArg.startsWith('/') ? inputArg : join(ROOT, inputArg);
if (!existsSync(inputPath)) {
  console.error(`File not found: ${inputPath}`);
  process.exit(1);
}

const data       = JSON.parse(readFileSync(inputPath, 'utf8'));
const candidates = (data.candidates ?? []).filter(c => !c._raw);
const doi        = data.doi ?? null;
const title      = data.title ?? null;
const minedAt    = data.mined_at ?? new Date().toISOString();

if (!candidates.length) {
  console.log('No candidates found in this file.');
  process.exit(0);
}

const slug     = basename(inputPath, extname(inputPath));
const outBase  = join(EXP_DIR, slug);
mkdirSync(EXP_DIR, { recursive: true });

console.log(`export-candidates — ${candidates.length} candidate(s) from ${basename(inputPath)}\n`);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// Ontology namespace → GitHub repo
const GITHUB_REPOS = {
  PO:    'https://github.com/Planteome/plant-ontology/issues/new',
  TO:    'https://github.com/Planteome/plant-trait-ontology/issues/new',
  PECO:  'https://github.com/Planteome/plant-experimental-conditions-ontology/issues/new',
  FLOPO: 'https://github.com/flora-phenotype-ontology/flopoontology/issues/new',
};

// Namespace → OBO ontology prefix
const NAMESPACE_PREFIX = {
  plant_anatomy:             'PO',
  plant_morphology:          'PO',
  plant_developmental_stage: 'PO',
  trait:                     'TO',
};

function termPrefix(c) {
  if (NAMESPACE_PREFIX[c.namespace]) return NAMESPACE_PREFIX[c.namespace];
  if (c.ontology === 'TO') return 'TO';
  if (c.ontology === 'PO') return 'PO';
  return null; // 'either' or unknown — curator must assign namespace
}

// Placeholder ID — curators replace with real accession
function placeholderId(c, index) {
  const prefix = termPrefix(c);
  const num    = String(index + 1).padStart(3, '0');
  return prefix ? `${prefix}:NEWTERM_${num}` : `??:NEWTERM_${num}`;
}

function doiRef(d) {
  return d ? `DOI:${d}` : 'REF:UNKNOWN';
}

function escapeOboDef(s) {
  return s.replace(/"/g, '\\"');
}

// ---------------------------------------------------------------------------
// Format: OBO stanzas
// ---------------------------------------------------------------------------

function buildObo() {
  const lines = [
    `! OBO stanza export`,
    `! Source   : ${inputPath}`,
    `! Paper    : ${title ?? '(unknown)'}`,
    `! DOI      : ${doi ?? '(unknown)'}`,
    `! Mined at : ${minedAt}`,
    `! Note     : Replace NEWTERM_XXX IDs with real accessions before committing.`,
    `!            Review and edit definition_draft before committing.`,
    '',
  ];

  for (const [i, c] of candidates.entries()) {
    const id = placeholderId(c, i);
    lines.push(`[Term]`);
    lines.push(`id: ${id}`);
    lines.push(`name: ${c.term}`);
    if (c.namespace) lines.push(`namespace: ${c.namespace}`);
    if (c.definition_draft) lines.push(`def: "${escapeOboDef(c.definition_draft)}" [${doiRef(doi)}]`);
    if (c.suggested_parent) lines.push(`! suggested_parent: ${c.suggested_parent}  (verify PO/TO accession before using is_a)`);
    for (const syn of c.synonyms ?? []) lines.push(`synonym: "${syn}" EXACT []`);
    if (c.source_sentence) lines.push(`comment: Source: "${c.source_sentence.slice(0, 200).replace(/"/g, '\\"')}"`);
    lines.push(`! confidence: ${c.confidence}`);
    lines.push('');
  }

  return lines.join('\n');
}

// ---------------------------------------------------------------------------
// Format: ROBOT template TSV
// ---------------------------------------------------------------------------

function buildRobot() {
  // ROBOT template header: https://robot.obolibrary.org/template
  const headers = ['ID', 'Label', 'Definition', 'Parent', 'Exact Synonym', 'Comment'];
  const robotHeaders = [
    'ID',
    'LABEL',
    "A definition 'definition'@en",
    "SC %",                          // subClassOf parent
    "A oboInOwl:hasExactSynonym",
    "A rdfs:comment",
  ];

  const rows = [headers, robotHeaders];

  for (const [i, c] of candidates.entries()) {
    const id = placeholderId(c, i);
    rows.push([
      id,
      c.term,
      c.definition_draft ?? '',
      c.suggested_parent ?? '',
      (c.synonyms ?? []).join('|'),
      [
        c.rationale ?? '',
        doi ? `Source DOI: ${doi}` : '',
        `Confidence: ${c.confidence}`,
        c.source_sentence ? `Evidence: "${c.source_sentence.slice(0, 150)}"` : '',
      ].filter(Boolean).join(' | '),
    ]);
  }

  return rows.map(r => r.map(cell => `"${String(cell).replace(/"/g, '""')}"`).join('\t')).join('\n');
}

// ---------------------------------------------------------------------------
// Format: GitHub NTR issues (one markdown file per term, grouped by ontology)
// ---------------------------------------------------------------------------

function buildGithubIssue(c) {
  const isAmbiguous = !c.ontology || c.ontology === 'either';
  const onto  = c.ontology === 'TO' ? 'TO' : c.ontology === 'PO' ? 'PO' : null;
  const repo  = onto ? GITHUB_REPOS[onto] : null;

  const lines = [
    `## New Term Request — ${onto ?? 'PO or TO (curator: please decide)'}`,
    '',
    `**Proposed label:** ${c.term}`,
    onto ? `**Ontology:** ${onto === 'PO' ? 'Plant Ontology (PO)' : 'Trait Ontology (TO)'}` : `**Ontology:** ⚠ Ambiguous — curator must determine whether this belongs in PO or TO`,
    c.namespace ? `**Namespace:** \`${c.namespace}\`` : null,
    '',
    '### Definition',
    '',
    c.definition_draft ?? '_(draft — please review)_',
    '',
    '### Suggested parent term',
    '',
    c.suggested_parent ? `\`${c.suggested_parent}\`` : '_(see rationale)_',
    '',
  ];

  if (c.synonyms?.length) {
    lines.push('### Synonyms', '');
    for (const s of c.synonyms) lines.push(`- \`${s}\` (EXACT)`);
    lines.push('');
  }

  lines.push(
    '### Evidence',
    '',
    doi   ? `**Paper DOI:** ${doi}` : null,
    title ? `**Paper title:** ${title}` : null,
    '',
    '**Supporting sentence from paper:**',
    '',
    `> ${c.source_sentence ?? '_(not available)_'}`,
    '',
    '### Rationale',
    '',
    c.rationale ?? '_(auto-generated candidate — please review)_',
    '',
    '### Automated check (OLS)',
    '',
    c.ols_search_result
      ? `OLS search result: ${c.ols_search_result}`
      : '_OLS check not recorded_',
    '',
    '---',
    `_Candidate generated by [llm-ontology-miner](https://github.com/jdetras/llm-ontology-miner) · confidence: ${c.confidence}_`,
    '',
    isAmbiguous
      ? `**Curator note:** Ontology is ambiguous — decide before submitting.\n- [PO Issues](${GITHUB_REPOS.PO})\n- [TO Issues](${GITHUB_REPOS.TO})`
      : `**Submit to:** [${onto} GitHub Issues](${repo})`,
  );

  return lines.filter(l => l !== null).join('\n');
}

function buildGithubAll() {
  return candidates.map((c, i) => `<!-- TERM ${i + 1} of ${candidates.length} -->\n\n${buildGithubIssue(c)}`).join('\n\n---\n\n');
}

// ---------------------------------------------------------------------------
// Format: CSV (human review)
// ---------------------------------------------------------------------------

function csvCell(v) {
  const s = String(v ?? '').replace(/"/g, '""');
  return `"${s}"`;
}

function buildCsv() {
  const headers = ['#', 'confidence', 'ontology', 'namespace', 'term', 'definition_draft', 'suggested_parent', 'synonyms', 'source_sentence', 'rationale', 'ols_check', 'doi'];
  const rows = [headers.map(csvCell).join(',')];
  for (const [i, c] of candidates.entries()) {
    rows.push([
      i + 1,
      c.confidence,
      c.ontology,
      c.namespace,
      c.term,
      c.definition_draft,
      c.suggested_parent,
      (c.synonyms ?? []).join('; '),
      c.source_sentence,
      c.rationale,
      c.ols_search_result ?? '',
      doi ?? '',
    ].map(csvCell).join(','));
  }
  return rows.join('\n');
}

// ---------------------------------------------------------------------------
// Write outputs
// ---------------------------------------------------------------------------

const formats = formatArg === 'all' ? ['obo', 'robot', 'github', 'csv'] : [formatArg];
const written = [];

for (const fmt of formats) {
  let content, ext, label;
  switch (fmt) {
    case 'obo':
      content = buildObo();       ext = '.obo';      label = 'OBO stanzas';
      break;
    case 'robot':
      content = buildRobot();     ext = '-robot.tsv'; label = 'ROBOT template TSV';
      break;
    case 'github':
      content = buildGithubAll(); ext = '-ntr.md';    label = 'GitHub NTR markdown';
      break;
    case 'csv':
      content = buildCsv();       ext = '.csv';       label = 'CSV for review';
      break;
    default:
      console.warn(`Unknown format: ${fmt} (use obo | robot | github | csv | all)`);
      continue;
  }
  const outPath = `${outBase}${ext}`;
  writeFileSync(outPath, content, 'utf8');
  written.push({ label, path: outPath });
  console.log(`✓ ${label.padEnd(22)} → ontologies/exports/${basename(outPath)}`);
}

console.log(`\n${candidates.length} candidate(s) exported in ${written.length} format(s).`);

// Print submission links for GitHub NTR
if (formats.includes('github')) {
  const ontos = [...new Set(candidates.flatMap(c =>
    c.ontology === 'TO' ? ['TO'] : c.ontology === 'PO' ? ['PO'] : ['PO', 'TO']
  ))];
  console.log('\nSubmission links:');
  for (const onto of ontos) {
    const repo = GITHUB_REPOS[onto];
    if (repo) console.log(`  ${onto}: ${repo}`);
  }
  const ambiguous = candidates.filter(c => !c.ontology || c.ontology === 'either');
  if (ambiguous.length) console.log(`\n  ⚠ ${ambiguous.length} candidate(s) have ambiguous ontology — see curator note in the NTR file`);
  console.log('\nTip: open the -ntr.md file, copy each term block, and paste into a new GitHub issue.');
}

if (formats.includes('robot')) {
  console.log('\nROBOT import command:');
  console.log(`  robot template --input your-ontology.owl --template ${outBase}-robot.tsv --output new-terms.owl`);
}

File: export_candidates.py
#!/usr/bin/env python3
"""
export_candidates.py — Convert mined candidates to ontology submission formats

Reads a candidates JSON file produced by ontology_miner.py or ontology_agent.py
and exports submission-ready files for ontology curators.

Output formats:
  obo      — OBO stanza blocks, ready to paste into a .obo file or GitHub PR
  robot    — ROBOT template TSV, importable with `robot template`
  github   — GitHub New Term Request (NTR) markdown, one file per term
  csv      — Flat CSV for spreadsheet-based human review
  all      — All of the above (default)

Usage:
    python export_candidates.py --input ontologies/candidates/my-paper-2026-06-05.json
    python export_candidates.py --input ontologies/candidates/my-paper-2026-06-05.json --format obo
    python export_candidates.py --list
"""

import argparse
import csv
import io
import json
import sys
from pathlib import Path

ROOT     = Path(__file__).parent
CAND_DIR = ROOT / "ontologies" / "candidates"
EXP_DIR  = ROOT / "ontologies" / "exports"

# ---------------------------------------------------------------------------
# Arg parsing
# ---------------------------------------------------------------------------

def parse_args():
    p = argparse.ArgumentParser(description="Export candidate ontology terms to submission formats")
    p.add_argument("--input",  help="Path to a candidates JSON file")
    p.add_argument("--format", default="all", choices=["obo", "robot", "github", "csv", "all"])
    p.add_argument("--list",   action="store_true", help="List available candidate files and exit")
    return p.parse_args()

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

GITHUB_REPOS = {
    "PO":    "https://github.com/Planteome/plant-ontology/issues/new",
    "TO":    "https://github.com/Planteome/plant-trait-ontology/issues/new",
    "PECO":  "https://github.com/Planteome/plant-experimental-conditions-ontology/issues/new",
    "FLOPO": "https://github.com/flora-phenotype-ontology/flopoontology/issues/new",
}

NAMESPACE_PREFIX = {
    "plant_anatomy":             "PO",
    "plant_morphology":          "PO",
    "plant_developmental_stage": "PO",
    "trait":                     "TO",
}


def term_prefix(c: dict) -> str | None:
    ns = NAMESPACE_PREFIX.get(c.get("namespace", ""))
    if ns:
        return ns
    if c.get("ontology") == "TO":
        return "TO"
    if c.get("ontology") == "PO":
        return "PO"
    return None  # 'either' or unknown — curator must assign namespace


def placeholder_id(c: dict, index: int) -> str:
    prefix = term_prefix(c)
    num    = str(index + 1).zfill(3)
    return f"{prefix}:NEWTERM_{num}" if prefix else f"??:NEWTERM_{num}"


def doi_ref(doi: str | None) -> str:
    return f"DOI:{doi}" if doi else "REF:UNKNOWN"

# ---------------------------------------------------------------------------
# Format: OBO stanzas
# ---------------------------------------------------------------------------

def build_obo(candidates: list, doi: str | None, title: str | None, mined_at: str, input_path: Path) -> str:
    lines = [
        "! OBO stanza export",
        f"! Source   : {input_path}",
        f"! Paper    : {title or '(unknown)'}",
        f"! DOI      : {doi or '(unknown)'}",
        f"! Mined at : {mined_at}",
        "! Note     : Replace NEWTERM_XXX IDs with real accessions before committing.",
        "!            Review and edit definition_draft before committing.",
        "",
    ]
    for i, c in enumerate(candidates):
        pid = placeholder_id(c, i)
        lines.append("[Term]")
        lines.append(f"id: {pid}")
        lines.append(f"name: {c['term']}")
        if c.get("namespace"):
            lines.append(f"namespace: {c['namespace']}")
        if c.get("definition_draft"):
            defn = c["definition_draft"].replace('"', '\\"')
            lines.append(f'def: "{defn}" [{doi_ref(doi)}]')
        if c.get("suggested_parent"):
            lines.append(f"! suggested_parent: {c['suggested_parent']}  (verify PO/TO accession before using is_a)")
        for syn in c.get("synonyms") or []:
            lines.append(f'synonym: "{syn}" EXACT []')
        if c.get("source_sentence"):
            src = c["source_sentence"][:200].replace('"', '\\"')
            lines.append(f'comment: Source: "{src}"')
        lines.append(f"! confidence: {c.get('confidence', 'unknown')}")
        lines.append("")
    return "\n".join(lines)

# ---------------------------------------------------------------------------
# Format: ROBOT template TSV
# ---------------------------------------------------------------------------

def build_robot(candidates: list, doi: str | None) -> str:
    headers      = ["ID", "Label", "Definition", "Parent", "Exact Synonym", "Comment"]
    robot_header = ["ID", "LABEL", "A definition 'definition'@en", "SC %", "A oboInOwl:hasExactSynonym", "A rdfs:comment"]

    buf = io.StringIO()
    w   = csv.writer(buf, delimiter="\t", quotechar='"', quoting=csv.QUOTE_ALL)
    w.writerow(headers)
    w.writerow(robot_header)

    for i, c in enumerate(candidates):
        comment_parts = [
            c.get("rationale") or "",
            f"Source DOI: {doi}" if doi else "",
            f"Confidence: {c.get('confidence', 'unknown')}",
            f"Evidence: \"{c['source_sentence'][:150]}\"" if c.get("source_sentence") else "",
        ]
        w.writerow([
            placeholder_id(c, i),
            c.get("term", ""),
            c.get("definition_draft", ""),
            c.get("suggested_parent", ""),
            "|".join(c.get("synonyms") or []),
            " | ".join(p for p in comment_parts if p),
        ])

    return buf.getvalue()

# ---------------------------------------------------------------------------
# Format: GitHub NTR issues
# ---------------------------------------------------------------------------

def build_github_issue(c: dict, doi: str | None, title: str | None) -> str:
    is_ambiguous = not c.get("ontology") or c.get("ontology") == "either"
    onto  = c.get("ontology") if c.get("ontology") in ("PO", "TO") else None
    repo  = GITHUB_REPOS.get(onto) if onto else None

    onto_label = onto or "PO or TO (curator: please decide)"
    if onto == "PO":
        onto_full = "Plant Ontology (PO)"
    elif onto == "TO":
        onto_full = "Trait Ontology (TO)"
    else:
        onto_full = None

    lines = [
        f"## New Term Request — {onto_label}",
        "",
        f"**Proposed label:** {c.get('term', '')}",
        f"**Ontology:** {onto_full}" if onto_full else "**Ontology:** ⚠ Ambiguous — curator must determine whether this belongs in PO or TO",
        f"**Namespace:** `{c['namespace']}`" if c.get("namespace") else None,
        "",
        "### Definition",
        "",
        c.get("definition_draft") or "_(draft — please review)_",
        "",
        "### Suggested parent term",
        "",
        f"`{c['suggested_parent']}`" if c.get("suggested_parent") else "_(see rationale)_",
        "",
    ]

    if c.get("synonyms"):
        lines += ["### Synonyms", ""]
        for s in c["synonyms"]:
            lines.append(f"- `{s}` (EXACT)")
        lines.append("")

    lines += [
        "### Evidence",
        "",
        f"**Paper DOI:** {doi}" if doi else None,
        f"**Paper title:** {title}" if title else None,
        "",
        "**Supporting sentence from paper:**",
        "",
        f"> {c.get('source_sentence') or '_(not available)_'}",
        "",
        "### Rationale",
        "",
        c.get("rationale") or "_(auto-generated candidate — please review)_",
        "",
        "### Automated check (OLS)",
        "",
        f"OLS search result: {c['ols_search_result']}" if c.get("ols_search_result") else "_OLS check not recorded_",
        "",
        "---",
        f"_Candidate generated by [llm-ontology-miner](https://github.com/jdetras/llm-ontology-miner) · confidence: {c.get('confidence', 'unknown')}_",
        "",
        (
            f"**Curator note:** Ontology is ambiguous — decide before submitting.\n- [PO Issues]({GITHUB_REPOS['PO']})\n- [TO Issues]({GITHUB_REPOS['TO']})"
            if is_ambiguous
            else f"**Submit to:** [{onto} GitHub Issues]({repo})"
        ),
    ]

    return "\n".join(line for line in lines if line is not None)


def build_github_all(candidates: list, doi: str | None, title: str | None) -> str:
    parts = [f"<!-- TERM {i + 1} of {len(candidates)} -->\n\n{build_github_issue(c, doi, title)}" for i, c in enumerate(candidates)]
    return "\n\n---\n\n".join(parts)

# ---------------------------------------------------------------------------
# Format: CSV (human review)
# ---------------------------------------------------------------------------

def build_csv(candidates: list, doi: str | None) -> str:
    fieldnames = ["#", "confidence", "ontology", "namespace", "term", "definition_draft",
                  "suggested_parent", "synonyms", "source_sentence", "rationale", "ols_check", "doi"]
    buf = io.StringIO()
    w   = csv.DictWriter(buf, fieldnames=fieldnames, quoting=csv.QUOTE_ALL)
    w.writeheader()
    for i, c in enumerate(candidates):
        w.writerow({
            "#":               i + 1,
            "confidence":      c.get("confidence", ""),
            "ontology":        c.get("ontology", ""),
            "namespace":       c.get("namespace", ""),
            "term":            c.get("term", ""),
            "definition_draft":c.get("definition_draft", ""),
            "suggested_parent":c.get("suggested_parent", ""),
            "synonyms":        "; ".join(c.get("synonyms") or []),
            "source_sentence": c.get("source_sentence", ""),
            "rationale":       c.get("rationale", ""),
            "ols_check":       c.get("ols_search_result", ""),
            "doi":             doi or "",
        })
    return buf.getvalue()

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    args = parse_args()

    if args.list:
        if not CAND_DIR.exists():
            print("No candidates directory found.")
            return
        files = sorted([f for f in CAND_DIR.iterdir() if f.suffix == ".json" and f.name != ".gitkeep"], reverse=True)
        if not files:
            print("No candidate files found in ontologies/candidates/")
            return
        print("Candidate files in ontologies/candidates/:\n")
        for f in files:
            data = json.loads(f.read_text())
            print(f"  {f.name}  ({data.get('candidate_count', 0)} candidates)")
        return

    if not args.input:
        print("Usage: python export_candidates.py --input <candidates-json> [--format obo|robot|github|csv|all]", file=sys.stderr)
        print("       python export_candidates.py --list", file=sys.stderr)
        sys.exit(1)

    input_path = Path(args.input) if Path(args.input).is_absolute() else ROOT / args.input
    if not input_path.exists():
        print(f"File not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    data       = json.loads(input_path.read_text())
    candidates = [c for c in (data.get("candidates") or []) if not c.get("_raw")]
    doi        = data.get("doi")
    title      = data.get("title")
    mined_at   = data.get("mined_at", "")

    if not candidates:
        print("No candidates found in this file.")
        return

    slug     = input_path.stem
    out_base = EXP_DIR / slug
    EXP_DIR.mkdir(parents=True, exist_ok=True)

    print(f"export-candidates — {len(candidates)} candidate(s) from {input_path.name}\n")

    formats = ["obo", "robot", "github", "csv"] if args.format == "all" else [args.format]
    written = []

    for fmt in formats:
        if fmt == "obo":
            content  = build_obo(candidates, doi, title, mined_at, input_path)
            ext, label = ".obo", "OBO stanzas"
        elif fmt == "robot":
            content  = build_robot(candidates, doi)
            ext, label = "-robot.tsv", "ROBOT template TSV"
        elif fmt == "github":
            content  = build_github_all(candidates, doi, title)
            ext, label = "-ntr.md", "GitHub NTR markdown"
        elif fmt == "csv":
            content  = build_csv(candidates, doi)
            ext, label = ".csv", "CSV for review"
        else:
            print(f"Unknown format: {fmt}", file=sys.stderr)
            continue

        out_path = out_base.parent / f"{out_base.name}{ext}"
        out_path.write_text(content)
        written.append((label, out_path))
        print(f"✓ {label:<22} → ontologies/exports/{out_path.name}")

    print(f"\n{len(candidates)} candidate(s) exported in {len(written)} format(s).")

    if "github" in formats:
        ontos = sorted({
            onto
            for c in candidates
            for onto in (["PO", "TO"] if not c.get("ontology") or c.get("ontology") == "either" else [c["ontology"]])
            if onto in GITHUB_REPOS
        })
        print("\nSubmission links:")
        for onto in ontos:
            print(f"  {onto}: {GITHUB_REPOS[onto]}")
        ambiguous = [c for c in candidates if not c.get("ontology") or c.get("ontology") == "either"]
        if ambiguous:
            print(f"\n  ⚠ {len(ambiguous)} candidate(s) have ambiguous ontology — see curator note in the NTR file")
        print("\nTip: open the -ntr.md file, copy each term block, and paste into a new GitHub issue.")

    if "robot" in formats:
        print("\nROBOT import command:")
        print(f"  robot template --input your-ontology.owl --template {out_base}-robot.tsv --output new-terms.owl")


if __name__ == "__main__":
    main()
